graph/path/shortest.go

```diff
@@ -174,6 +174,250 @@ func (p Shortest) To(vid int64) (path []graph.Node, weight float64) {
+// a negative cycle, the returned weight will not reflect the true path weight.
+func (p ShortestAlts) WeightTo(vid int64) float64 {
+	to, toOK := p.indexOf[vid]
+	if !toOK {
+		return math.Inf(1)
+	}
+	return p.dist[to]
+}
+
+// To returns a shortest path to v and the weight of the path. If more than
+// one shortest path exists between u and v, a randomly chosen path will be
+// returned and unique is returned false. If a cycle with zero weight exists
+// in the path, it will not be included, but unique will be returned false.
+// If the path to v includes a negative cycle, one pass through the cycle will
+// be included in path, but any path leading into the negative cycle will be
+// lost, and weight will be returned as -Inf.
+func (p ShortestAlts) To(vid int64) (path []graph.Node, weight float64, unique bool) {
+	to, toOK := p.indexOf[vid]
+	if !toOK || math.IsInf(p.dist[to], 1) {
+		return nil, math.Inf(1), false
+	}
+	from := p.indexOf[p.from.ID()]
+	unique = true
+	path = []graph.Node{p.nodes[to]}
+	if p.hasNegativeCycle {
+		weight = math.Inf(1)
+		seen := make(set.Ints)
+		seen.Add(from)
+		for to != from {
+			c := p.next[to]
+			var next int
+			if len(c) != 1 {
+				unique = false
+				next = c[rand.Intn(len(c))]
+			} else {
+				next = c[0]
+			}
+			if math.IsInf(p.cycCosts[[2]int{next, to}], -1) {
+				weight = math.Inf(-1)
+				unique = false
+			}
+			if seen.Has(to) {
+				break
+			}
+			seen.Add(to)
+			path = append(path, p.nodes[next])
+			to = next
+		}
+		weight = math.Min(weight, p.dist[p.indexOf[vid]])
+	} else {
+		seen := make([]int, len(p.nodes))
+		for i := range seen {
+			seen[i] = -1
+		}
+		seen[to] = 0
+
+		var next int
+		for from != to {
+			c := p.next[to]
+			if len(c) != 1 {
+				unique = false
+				next = c[rand.Intn(len(c))]
+			} else {
+				next = c[0]
+			}
+			if seen[next] >= 0 {
+				path = path[:seen[next]]
+			}
+			seen[next] = len(path)
+			path = append(path, p.nodes[next])
+			to = next
+		}
+		weight = p.dist[p.indexOf[vid]]
+	}
+
+	ordered.Reverse(path)
+	return path, weight, unique
+}
+
+// AllTo returns all shortest paths to v and the weight of the paths. Paths
+// containing zero-weight cycles are not returned. If a negative cycle exists between
+// u and v, paths is returned nil and weight is returned as -Inf.
+func (p ShortestAlts) AllTo(vid int64) (paths [][]graph.Node, weight float64) {
+	from := p.indexOf[p.from.ID()]
+	to, toOK := p.indexOf[vid]
+	if !toOK || len(p.next[to]) == 0 {
+		if p.from.ID() == vid {
+			return [][]graph.Node{{p.nodes[from]}}, 0
+		}
+		return nil, math.Inf(1)
+	}
+
+	_, weight, unique := p.To(vid)
+	if math.IsInf(weight, -1) && !unique {
+		return nil, math.Inf(-1)
+	}
+
+	seen := make([]bool, len(p.nodes))
+	paths = p.allTo(from, to, seen, []graph.Node{p.nodes[to]}, nil)
+	weight = p.dist[to]
+
+	return paths, weight
+}
+
+func (p ShortestAlts) allTo(from, to int, seen []bool, path []graph.Node, paths [][]graph.Node) [][]graph.Node {
```

Comment:
This function is long enough to deserve a descriptive comment.

Reply:
I have made an issue to address these comments #1388; they exist on all three types, so I would prefer to have a single commit adding to all, distinct from this PR which is logic bearing.

Reply:
#1389 is the draft, which I'll update for this PR's inclusions when it's merged.